src/main/cocoa/CocoaWindow.mm

```diff
@@ -0,0 +1,731 @@
+            {
+                if (pCocoaWindow == NULL)
+                {
+                    nActions = actions;
+                    return STATUS_OK;
+                }
+
+                return (nActions != actions) ? commit_border_style(enBorderStyle, actions) : STATUS_OK;
+            }
+
+            status_t CocoaWindow::get_window_actions(size_t *actions)
+            {
+                if (actions == NULL)
+                    return STATUS_BAD_ARGUMENTS;
+                if (pCocoaWindow == NULL)
+                    return STATUS_BAD_STATE;
+
+                *actions    = nActions;
+                return STATUS_OK;
+            }
+
+            status_t CocoaWindow::resize(ssize_t width, ssize_t height) {
+                rectangle_t rect = sSize;
+                rect.nWidth = width;
+                rect.nHeight = height;
+                return set_geometry(&rect);
+            }
+
+            void CocoaWindow::apply_constraints(rectangle_t *dst, const rectangle_t *req) {
+                *dst = *req;
+
+                if ((sConstraints.nMaxWidth >= 0) && (dst->nWidth > sConstraints.nMaxWidth))
+                    dst->nWidth = sConstraints.nMaxWidth;
+                if ((sConstraints.nMaxHeight >= 0) && (dst->nHeight > sConstraints.nMaxHeight))
+                    dst->nHeight = sConstraints.nMaxHeight;
+                if ((sConstraints.nMinWidth >= 0) && (dst->nWidth < sConstraints.nMinWidth))
+                    dst->nWidth = sConstraints.nMinWidth;
+                if ((sConstraints.nMinHeight >= 0) && (dst->nHeight < sConstraints.nMinHeight))
+                    dst->nHeight = sConstraints.nMinHeight;
+            }
+
+            status_t CocoaWindow::set_geometry(const rectangle_t *realize) {
+                if (!pCocoaWindow)
+                    return STATUS_BAD_STATE;
+
+                rectangle_t old = sSize;
+                apply_constraints(&sSize, realize);
+
+                if ((old.nLeft == sSize.nLeft) &&
+                    (old.nTop == sSize.nTop) &&
+                    (old.nWidth == sSize.nWidth) &&
+                    (old.nHeight == sSize.nHeight))
+                        return STATUS_OK;
+                
+                return set_geometry_impl();
+            }
+
+            status_t CocoaWindow::set_geometry_impl() {
+                if (!pCocoaWindow)
+                    return STATUS_BAD_STATE;
+
+                // Calculate the frame rect from the content rect
+                printf("Resize / move window {nL=%d, nT=%d, nW=%d, nH=%d}\n", int(sSize.nLeft), int(sSize.nTop), int(sSize.nWidth), int(sSize.nHeight));
+                NSRect contentRect = NSMakeRect(sSize.nLeft, sSize.nTop, sSize.nWidth, sSize.nHeight);
+                NSRect frameRect = [pCocoaWindow frameRectForContentRect:contentRect];
+
+                
+                // macOS coordinates start from bottom-left, so we may need to flip Y if using a fixed origin
```

Comment:
Yes, we assume the normal screen coordinate system where the origin is located in the left topmost pixel.

Reply:
Origin on window creation 0x0 should also left top?